### PR TITLE
Added span.hpp from libc++. Not sure how to test it, and I'm not sure…

### LINKED ISSUE #577: Import std::span from libc++ into our std-module
Copy'n'aste from:
https://github.com/llvm-mirror/libcxx/blob/master/include/span

* include copyright holder in the header
* just remove `_LIBCPP_TEMPLATE_VIS` and `_LIBCPP_INLINE_VISIBILITY`
* adapt to our code-style an make sure stuff is not named `__foo`, but just `foo`
* don't import the comparison operators, because they are deprecated for span
* instead of `_LIBCPP_STD_VER > 17` check `__cplusplus <= 201703L`

### PR DESCRIPTION
**Pull request by joshuak94**
resolves #577

---

… what to do about the comments/documentation currently there...

### CODE REVIEW

**Comment by marehr** [comment]
Something does not work yet, the header tests are failing:

```c++
/home/travis/build/seqan/seqan3/include/seqan3/std/span.hpp:11:0: error: unterminated #ifndef
 #ifndef _LIBCPP_SPAN
```

and documentation is broken, because nothing is commented

File: include/seqan3/std/span.hpp [hunk 1]
```diff
@@ -0,0 +1,551 @@
...
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is dual licensed under the MIT and the University of Illinois Open
+// Source Licenses. See LICENSE.TXT for details.
+//
+//===---------------------------------------------------------------------===//
+
+#ifndef _LIBCPP_SPAN
+#define _LIBCPP_SPAN
```
Comment (marehr) [inline include/seqan3/std/span.hpp]: replace header guard with 
```suggestion
#pragma once
```

File: include/seqan3/std/span.hpp [hunk 2]
```diff
@@ -0,0 +1,551 @@
...
+
```
Comment (marehr) [inline include/seqan3/std/span.hpp]: add
```suggestion
//!\cond
```

to disable documentation stuff

File: include/seqan3/std/span.hpp [hunk 3]
```diff
@@ -0,0 +1,551 @@
...
+    span(const array<span_tp, span_sz>&) -> span<const span_tp, span_sz>;
+
+template<class container>
+    span(container&) -> span<typename container::value_type>;
+
+template<class container>
+    span(const container&) -> span<const typename container::value_type>;
+} // namespace std
+
+#endif // __cplusplus <= 201703L
```
Comment (marehr) [inline include/seqan3/std/span.hpp]: and close the no-documentation scope

```suggestion
#endif // __cplusplus <= 201703L
//!\endcond
```

**Comment by marehr** [comment]
We decided that STL imported code does not need documentation. But I'm not sure with regards to tests. I'm inclined to say we don't need it, but I would want to clarify it with the core team first.

**Comment by smehringer** [comment]
IMPOV I do think we should have a small test for it. Nothing extensive or thorough but the basic usage.

**Comment by h-2** [comment]
> We decided that STL imported code does not need documentation. 

I think we decided on referencing the standard documentation, i.e.
```doxygen
\brief Provides std::foobar from the C++20 standard library.
\see https://en.cppreference.com/...
```
And then wrap the rest (all members and related functions) in `\cond` and `\endcond`.

> But I'm not sure with regards to tests. I'm inclined to say we don't need it, but I would want to clarify it with the core team first.

Well, I think for small things we don't need tests. For something as big as span (and since we did some renaming of variables and also smaller changes), maybe at least a small functionality test.

File: include/seqan3/std/span.hpp [hunk 4]
```diff
@@ -0,0 +1,551 @@
...
+    : public true_type {};
+
+
+template <typename span_tp, ptrdiff_t span_extent>
+class span {
+public:
+//  constants and types
+    using element_type           = span_tp;
+    using value_type             = remove_cv_t<span_tp>;
+    using index_type             = ptrdiff_t;
```
Comment (h-2) [inline include/seqan3/std/span.hpp]: this has now been changed to `size_t` so it's unsigned. We should change it in our version, too.

File: include/seqan3/std/span.hpp [hunk 5]
```diff
@@ -0,0 +1,551 @@
...
+
+    span<byte, span_extent * sizeof(element_type)> as_writeable_bytes() const noexcept
+    { return {reinterpret_cast<byte *>(get_data()), size_bytes()}; }
+
+private:
+    pointer data;
+
+};
+
+
```
Comment (h-2) [inline include/seqan3/std/span.hpp]: double newline

File: include/seqan3/std/span.hpp [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+// #include <__config>
+#pragma once
+
+#include <cstddef>      // for ptrdiff_t
+#include <iterator>     // for iterators
+#include <array>        // for array
+#include <type_traits>  // for remove_cv, etc
+#include <cstddef>      // for byte
+
+#if __has_include(<span>)
```
[the remark this answers is not in the document]
Comment (smehringer) [reply include/seqan3/std/span.hpp]: And you only want the above includes (array, type_traits.. etc.) IF you need to define the span here, so why are they above the if definition?

File: include/seqan3/std/span.hpp [hunk 7]
```diff
@@ -0,0 +1,551 @@
...
+// #include <__config>
+#include <cstddef>      // for ptrdiff_t
+#include <iterator>     // for iterators
+#include <array>        // for array
+#include <type_traits>  // for remove_cv, etc
+#include <cstddef>      // for byte
+
+#pragma once
+
+#if __cplusplus <= 201703L
```
[the remark this answers is not in the document]
Comment (joshuak94) [reply include/seqan3/std/span.hpp]: I'm confused now, should I be checking `#if __has_include(<span>)` or `if !__has_include(<span>)`?

File: include/seqan3/std/span.hpp [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+//
+// This file is dual licensed under the MIT and the University of Illinois Open
+// Source Licenses. See LICENSE.TXT for details.
+//
+//===---------------------------------------------------------------------===//
+
+
+/*!\file
+// \brief Provides std::foobar from the C++20 standard library.
+// \see https://en.cppreference.com/...
```
Comment (h-2) [inline include/seqan3/std/span.hpp]: Actually put the link here!

File: include/seqan3/std/span.hpp [hunk 9]
```diff
@@ -0,0 +1,421 @@
...
+*/
+
+//!\cond
+// #include <__config>
```
Comment (h-2) [inline include/seqan3/std/span.hpp]: whats this?

File: include/seqan3/std/span.hpp [hunk 10]
```diff
@@ -0,0 +1,421 @@
...
+#pragma once
+
+#include <cstddef>      // for ptrdiff_t
+#include <iterator>     // for iterators
+#include <array>        // for array
+#include <type_traits>  // for remove_cv, etc
+#include <cstddef>      // for byte
```
Comment (h-2) [inline include/seqan3/std/span.hpp]: sort alphbatically

File: include/seqan3/std/span.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (h-2) [reply include/seqan3/std/span.hpp]: Yes, it should be 
```cpp
#if __has_include(<span>)
#include <span>
#else 
... rest of file
#endif
```

File: include/seqan3/std/span.hpp [hunk 11]
```diff
@@ -0,0 +1,421 @@
...
+
+template <typename span_tp>
+class span<span_tp, dynamic_extent> {
+private:
+
+public:
+//  constants and types
+    using element_type           = span_tp;
+    using value_type             = remove_cv_t<span_tp>;
+    using index_type             = ptrdiff_t;
```
Comment (h-2) [inline include/seqan3/std/span.hpp]: this should also be `size_t`

File: include/seqan3/std/span.hpp (hunk 9 above)
Comment (joshuak94) [reply include/seqan3/std/span.hpp]: Not sure, was in the original file and I didn't want to remove it, but removing it now!

**Pull request merged**